tool/teleport/common/sftp.go

```diff
@@ -75,17 +79,96 @@ func (c compositeCh) Close() error {
 	return trace.NewAggregate(c.r.Close(), c.w.Close())
 }
 
+type byteReader struct {
+	buf []byte
+	io.Reader
+}
+
+func newByteReader(r io.Reader) *byteReader {
+	return &byteReader{
+		Reader: r,
+		buf:    make([]byte, 1),
+	}
+}
+
+func (r *byteReader) ReadByte() (byte, error) {
+	_, err := r.Reader.Read(r.buf)
+	if err != nil {
+		return 0, err
+	}
+
+	return r.buf[0], nil
+}
+
+type allowedOps struct {
+	write bool
+	path  string
+}
+
 // sftpHandler provides handlers for a SFTP server.
 type sftpHandler struct {
-	logger *log.Entry
-	events chan<- *apievents.SFTP
+	logger  *log.Entry
+	allowed *allowedOps
+	events  chan<- *apievents.SFTP
 }
 
-func newSFTPHandler(logger *log.Entry, events chan<- *apievents.SFTP) *sftpHandler {
+func newSFTPHandler(logger *log.Entry, req *srv.FileTransferRequest, events chan<- *apievents.SFTP) (*sftpHandler, error) {
+	var allowed *allowedOps
+	if req != nil {
+		allowed = &allowedOps{
+			write: !req.Download,
+		}
+		// TODO(capnspacehook): reject relative paths and symlinks
+		// make filepaths consistent by ensuring all separators use backslashes
+		allowed.path = path.Clean(req.Location)
+	}
+
 	return &sftpHandler{
-		logger: logger,
-		events: events,
+		logger:  logger,
+		allowed: allowed,
+		events:  events,
+	}, nil
+}
+
+func newDisallowedErr(req *sftp.Request) error {
+	return fmt.Errorf("method %s is not allowed on %s", strings.ToLower(req.Method), req.Filepath)
+}
+
+// ensureReqIsAllowed returns an error if the SFTP request isn't
+// allowed based on the approved file transfer request for this session.
+func (s *sftpHandler) ensureReqIsAllowed(req *sftp.Request) error {
+	// no specifically allowed operations, all requests are allowed
+	if s.allowed == nil {
+		return nil
+	}
+
+	if s.allowed.path != path.Clean(req.Filepath) {
+		return newDisallowedErr(req)
+	}
+
+	switch req.Method {
+	case methodLstat, methodStat:
+		// these methods are allowed
+	case methodGet:
+		// only allow reads for downloads
+		if s.allowed.write {
+			return newDisallowedErr(req)
+		}
+	case methodPut:
+		// only allow writes for uploads
+		if !s.allowed.write {
+			return newDisallowedErr(req)
+		}
+	case methodSetStat:
+		// only allow chmods for uploads
+		if !s.allowed.write {
+			return newDisallowedErr(req)
+		}
```

Comment:
These 3 cases look identical? Can we combine them?

```
case methodGet, methodPut, methodSetStat:
```

Reply:
Yes good catch, they were different before but should be combined now.